File: 380_Insert_Delete_GetRandom_O(1)/ranset.rs
use rand::Rng;
use std::collections::HashMap;

struct RandomizedSet {
    index: HashMap<i32, usize>, // Maps values to their indices
    vals: Vec<i32>,             // Stores the values
}


/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl RandomizedSet {

    fn new() -> Self {
        RandomizedSet {
            index: HashMap::new(),
            vals: Vec::new(),
        }        
    }
    
    fn insert(&mut self, val: i32) -> bool {
        if self.index.contains_key(&val) {
            return false;
        }
        self.index.insert(val, self.vals.len());
        self.vals.push(val);
        true        
    }
    
    fn remove(&mut self, val: i32) -> bool {
        if let Some(&idx) = self.index.get(&val) {
            let last_val = self.vals.pop().unwrap();
            if idx != self.vals.len() {
                // Swap the value to remove with the last value
                self.vals[idx] = last_val;
                self.index.insert(last_val, idx);
            }
            self.index.remove(&val);
            true
        } else {
            false
        }        
    }
    
    fn get_random(&self) -> i32 {
        let mut rng = rand::thread_rng();
        let i = rng.gen_range(0..self.vals.len());
        self.vals[i]        
    }
}

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * let obj = RandomizedSet::new();
 * let ret_1: bool = obj.insert(val);
 * let ret_2: bool = obj.remove(val);
 * let ret_3: i32 = obj.get_random();
 */

File: problems/3/380 Insert Delete GetRandom O(1)/ranset.py
import random

class RandomizedSet:

    def __init__(self):
        self.index = dict()
        self.vals = list()        

    def insert(self, val: int) -> bool:
        if val in self.index:
            return False
        self.index[val] = len(self.vals)
        self.vals.append(val)
        return True

    def remove(self, val: int) -> bool:
        if val not in self.index:
            return False
        idx = self.index.pop(val)
        if idx == len(self.vals) - 1:
            self.vals.pop()
        else:
            self.vals[idx] = self.vals.pop()
            self.index[self.vals[idx]] = idx
        return True

    def getRandom(self) -> int:
        i = random.randint(0, len(self.vals) - 1)
        return self.vals[i]
        


# Your RandomizedSet object will be instantiated and called as such:
obj = RandomizedSet()
print(obj.insert(1))
print(obj.remove(2))
print(obj.insert(2))
print()

print(obj.getRandom())
print(obj.getRandom())
print(obj.getRandom())
print(obj.getRandom())
print(obj.getRandom())
print(obj.getRandom())
print()

print(obj.remove(1))
print(obj.insert(2))
print()

print(obj.getRandom())
print(obj.getRandom())
